File: day-20/rust/src/main.rs
use std::collections::{HashMap, VecDeque};

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum Pulse {
    High,
    Low,
}

#[derive(Debug, Clone)]
enum NodeKind<'a> {
    FlipFlop { on: bool },
    Conjunction { inputs: HashMap<&'a str, Pulse> },
    Broacaster,
}

#[derive(Debug, Clone)]
struct Node<'a> {
    name: &'a str,
    destinations: Vec<&'a str>,
    kind: NodeKind<'a>,
}

impl<'a> Node<'a> {
    fn parse(string: &'a str) -> Node<'a> {
        let (node, destinations) = string.split_once(" -> ").unwrap();

        let destinations = destinations.split(", ").collect();

        let (name, kind) = match node.chars().next().unwrap() {
            'b' => ("broadcaster", NodeKind::Broacaster),
            '%' => (&node[1..], NodeKind::FlipFlop { on: false }),
            '&' => (&node[1..], NodeKind::Conjunction { inputs: [].into() }),
            c => unreachable!("bad input: {c}"),
        };

        Self {
            name,
            destinations,
            kind,
        }
    }

    fn process(&mut self, source: &'a str, pulse: Pulse) -> Vec<Signal<'a>> {
        match &mut self.kind {
            NodeKind::FlipFlop { on } => {
                if pulse == Pulse::Low {
                    *on = !*on;
                    if *on {
                        self.destinations
                            .iter()
                            .map(|dest| Signal {
                                source: self.name,
                                dest,
                                pulse: Pulse::High,
                            })
                            .collect()
                    } else {
                        self.destinations
                            .iter()
                            .map(|dest| Signal {
                                source: self.name,
                                dest,
                                pulse: Pulse::Low,
                            })
                            .collect()
                    }
                } else {
                    vec![]
                }
            }
            NodeKind::Conjunction { inputs } => {
                let last = inputs.get_mut(source).unwrap();
                *last = pulse;
                if inputs.iter().all(|(_, pulse)| *pulse == Pulse::High) {
                    self.destinations
                        .iter()
                        .map(|dest| Signal {
                            source: self.name,
                            dest,
                            pulse: Pulse::Low,
                        })
                        .collect()
                } else {
                    self.destinations
                        .iter()
                        .map(|dest| Signal {
                            source: self.name,
                            dest,
                            pulse: Pulse::High,
                        })
                        .collect()
                }
            }
            NodeKind::Broacaster => todo!(),
        }
    }
}

struct Signal<'a> {
    source: &'a str,
    dest: &'a str,
    pulse: Pulse,
}

fn main() {
    // let input = include_str!("../../test_input.txt");
    // let input = include_str!("../../test_input2.txt");
    let input = include_str!("../../input.txt");

    let mut nodes: HashMap<&str, Node> = input
        .lines()
        .map(Node::parse)
        .map(|node| (node.name, node))
        .collect();

    // Maps a node name to its list of inputs.
    let mut input_map: HashMap<&str, Vec<&str>> = [].into();
    for node in nodes.values() {
        for destination in node.destinations.iter() {
            let inputs = input_map.entry(destination).or_default();
            inputs.push(node.name);
        }
    }

    // Update inputs for Conjunction nodes.
    for node in nodes.values_mut() {
        if let NodeKind::Conjunction { inputs, .. } = &mut node.kind {
            let ugh = input_map.get(node.name).unwrap();
            let ugh = ugh.iter().map(|input| (*input, Pulse::Low));
            inputs.extend(ugh);
        }
    }

    // dbg!(nodes);

    let mut queue: VecDeque<Signal> = [].into();
    let mut low_sent = 0;
    let mut high_sent = 0;

    let mut precursors: HashMap<&str, Option<u64>> =
        [("tf", None), ("db", None), ("vq", None), ("ln", None)].into();

    for button_press in 1.. {
        // Pressing the button also sends a Low pulse.
        low_sent += 1;

        // Start by broadcasting Low to all of broadcaster's destinations.
        let broadcaster = nodes.get("broadcaster").unwrap();
        for dest in broadcaster.destinations.iter() {
            queue.push_back(Signal {
                source: "broadcaster",
                dest,
                pulse: Pulse::Low,
            });
        }

        while let Some(Signal {
            source,
            dest,
            pulse,
        }) = queue.pop_front()
        {
            // Check if we're sending a pulse to one of rx's precursors.
            if pulse == Pulse::High {
                if let Some(precursor) = precursors.get_mut(source) {
                    precursor.get_or_insert(button_press);

                    // If we've calculated the loop size for all precurors,
                    // calculate their product (part 2) and exit.
                    if precursors.values().all(Option::is_some) {
                        let part2: u64 = precursors.into_values().map(|o| o.unwrap()).product();
                        println!("part2 = {part2}");
                        std::process::exit(0);
                    }
                }
            }

            match pulse {
                Pulse::High => high_sent += 1,
                Pulse::Low => low_sent += 1,
            };

            if dest == "output" {
                continue;
            }

            if let Some(node) = nodes.get_mut(dest) {
                let results = node.process(source, pulse);
                queue.extend(results);
            } else if pulse == Pulse::Low {
                println!("part2 = {}", button_press + 1);
            }
        }

        if button_press == 1000 {
            let part1 = low_sent * high_sent;
            println!("part1 = {part1}");
        }
    }
}

File: day-20/scala/src/main/scala/Main.scala
enum Pulse:
  case High
  case Low

enum NodeKind:
  case FlipFlop(on: Boolean)
  case Conjunction(inputs: collection.mutable.Map[String, Pulse])
  case Broadcaster

class Node(
    val name: String,
    val destinations: Array[String],
    val kind: NodeKind
):
  def with_kind(kind: NodeKind): Node = {
    Node(this.name, this.destinations, kind)
  }

  def process(source: String, pulse: Pulse): (Node, Array[Signal]) = {
    this.kind match
      case NodeKind.FlipFlop(on) => {
        pulse match
          case Pulse.Low => {
            val new_node = this.with_kind(NodeKind.FlipFlop(!on))
            val new_pulse =
              if !on
              then Pulse.High
              else Pulse.Low

            (
              new_node,
              this.destinations
                .map((dest) => Signal(this.name, dest, new_pulse))
                .toArray
            )
          }
          case Pulse.High => (this, Array())
      }
      case NodeKind.Conjunction(inputs) => {
        inputs.put(source, pulse)
        val new_pulse: Pulse =
          if inputs.values.forall(_ == Pulse.High)
          then Pulse.Low
          else Pulse.High

        (
          this,
          this.destinations
            .map((dest) => Signal(this.name, dest, new_pulse))
            .toArray
        )
      }
      case NodeKind.Broadcaster => throw new RuntimeException("unreachable?")
  }

def parse_node(line: String): Node = {
  val splitter = line.split(" -> ")
  val node = splitter(0)
  val destinations = splitter(1).split(", ")

  val (name, kind) = node(0) match
    case 'b' => ("broadcaster", NodeKind.Broadcaster)
    case '%' => (node.substring(1), NodeKind.FlipFlop(false))
    case '&' =>
      (node.substring(1), NodeKind.Conjunction(collection.mutable.Map()))

  Node(name, destinations, kind)
}

case class Signal(
    source: String,
    dest: String,
    pulse: Pulse
)

@main def main: Unit =
  // val input = io.Source.fromFile("../test_input.txt")
  // val input = io.Source.fromFile("../test_input2.txt")
  val input = io.Source.fromFile("../input.txt")

  val nodes = collection.mutable.Map[String, Node]()
  for (node <- input.getLines().map(parse_node)) {
    nodes.put(node.name, node)
  }

  val input_map = scala.collection.mutable.Map[String, List[String]]()
  for (node <- nodes.values) {
    for (destination <- node.destinations) {
      val inputs = input_map.getOrElse(destination, List[String]())
      input_map(destination) = node.name :: inputs
    }
  }

  // println(input_map)

  for (node <- nodes.values) {
    node.kind match
      case NodeKind.Conjunction(inputs) => {
        inputs ++= input_map.get(node.name).get.map((_, Pulse.Low))
      }
      case _ => ()
  }

  val queue = collection.mutable.Queue[Signal]()
  var low_sent = 0
  var high_sent = 0
  var button_press = 0

  val precursors = collection.mutable.Map[String, Option[Long]](
    ("tf", None),
    ("db", None),
    ("vq", None),
    ("ln", None)
  )

  while (true) {
    button_press += 1
    low_sent += 1

    val broadcaster = nodes.get("broadcaster").get
    queue.addAll(
      broadcaster.destinations.map((dest) =>
        Signal("broadcaster", dest, Pulse.Low)
      )
    )

    while (queue.nonEmpty) {
      val signal = queue.dequeue

      signal.pulse match
        case Pulse.High => {
          high_sent += 1
          precursors.get(signal.source) match
            case Some(Some(_existing)) => ()
            case Some(None) => precursors.put(signal.source, Some(button_press))
            case _          => ()
        }
        case Pulse.Low => low_sent += 1

      if precursors.values.forall(_.nonEmpty)
      then {
        val part2 = precursors.values.map(_.get).product
        println(s"part2 = $part2")
        System.exit(0)
      } else ()

      if signal.dest == "output"
      then ()
      else {
        nodes.get(signal.dest) match
          case Some(node) => {
            val (processed, result) = node.process(signal.source, signal.pulse)
            nodes.put(signal.dest, processed)
            queue.addAll(result)
          }
          case _ => () // TODO: part 2?
      }
    }

    if button_press == 1000
    then {
      val part1 = high_sent * low_sent
      println(s"part1 = $part1")
    }
  }
